janitor/functions/coalesce.py

```diff
@@ -93,14 +105,13 @@ def coalesce(
 
     if target_column_name is None:
         target_column_name = column_names[0]
-    # bfill/ffill combo is faster than combine_first
-    outcome = (
-        df.filter(column_names)
-        .bfill(axis="columns")
-        .ffill(axis="columns")
-        .iloc[:, 0]
-    )
+
+    outcome = df.filter(column_names).bfill(axis="columns").iloc[:, 0]
     if outcome.hasnans and (default_value is not None):
         outcome = outcome.fillna(default_value)
 
-    return df.assign(**{target_column_name: outcome})
+    # to allow for non-strings ... GH #1016
+    df = df.copy()
+    df[target_column_name] = outcome
```

Comment:
Actually, since we've "checked" that `target_column_name` is a string type a few lines above, technically #1016 shouldn't be a problem here.

So I guess the question here is more of whether `target_column_name` should be restricted to string-only or not. I have no issue keeping it status quo -- for "newcolumn names" that are being added by pyjanitor to be string-only.

(I'm somehow assuming the appearance of non-string hashable column names are because the data is being imported from somewhere, as opposed to someone explicitly wanting a non-string hashable column name; so I personally don't see the benefit of giving end-user the option to create non-string hashable named columns here.)